lib/msal-common/src/network/NetworkManager.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { ServerAuthorizationTokenResponse } from "../server/ServerAuthorizationTokenResponse";
+import { INetworkModule } from "./INetworkModule";
+import { ServerTokenRequestParameters } from "../server/ServerTokenRequestParameters";
+import { ICacheStorage } from "../cache/ICacheStorage";
+import { AccessTokenKey } from "../cache/AccessTokenKey";
+import { AccessTokenValue } from '../cache/AccessTokenValue';
+import { Constants } from "../utils/Constants";
+import { ICrypto } from "../crypto/ICrypto";
+import { TokenExchangeParameters } from "../request/TokenExchangeParameters";
+import { ClientInfo } from "../auth/ClientInfo";
+
+export type NetworkResponse<T> = {
+    headers: Map<string, string>;
+    body: T;
+    status: number;
+};
+
+export class NetworkManager {
+    cacheStorage: ICacheStorage;
+    networkClient: INetworkModule;
+    cryptoObj: ICrypto;
+
+    constructor(cacheStorage: ICacheStorage, networkClient: INetworkModule, cryptoObj: ICrypto) {
+        this.cacheStorage = cacheStorage;
+        this.networkClient = networkClient;
+        this.cryptoObj = cryptoObj;
+    }
+
+    async sendPostRequest(tokenEndpoint: string, tokenReqParams: ServerTokenRequestParameters, tokenRequest: TokenExchangeParameters, clientInfo: ClientInfo, clientId: string): Promise<ServerAuthorizationTokenResponse> {
+        const thumbprint = new AccessTokenKey(
+            tokenRequest.authority, 
+            clientId, 
+            tokenRequest.scopes.join(" "), 
+            tokenRequest.resource, 
+            clientInfo && clientInfo.uid, 
+            clientInfo && clientInfo.utid, 
+            this.cryptoObj
+        );
+
+        const options = {
+            body: tokenReqParams.createRequestBody(),
+            headers: tokenReqParams.createRequestHeaders()
+        };
+
+        // this.preProcess();
+
+        const networkResponse = await this.networkClient.sendPostRequestAsync(tokenEndpoint, options);
+
+        this.postProcess(networkResponse, thumbprint);
+
+        return networkResponse.body as ServerAuthorizationTokenResponse;
+    }
+
+    private postProcess(response: NetworkResponse<ServerAuthorizationTokenResponse>, thumbprint: AccessTokenKey): void {
+        if (response.status >= 500 || response.status == 429 || (response.status < 200 || response.status >= 300) && response.headers.has("Retry-After")) {
+            const throttleTime = parseInt(response.headers.get("Retry-After")) || Date.now() + Constants.DEFAULT_THROTTLE_TIME_MS;
```

Comment:
Don't you need to save the time at which the error response was received for this to have meaning? if I pull this item out of the cache, and it has a throttleTime property on it with related errors, what does that mean? how would that context "know" whether the throttle time has elapsed or not at the time the cache item is resolved? Or is the signal then for that cache item "observer" to wait for the specified number of seconds on top of however long the item has been cached? Or perhaps the acting of caching the item adds a cache dateTime property to it?

Reply:
I haven't explicitly tested this yet because I need to get some information about the field from the server team as I don't have a way of directly reproing these errors at this moment. That being said, my assumption was that they'd be passing us a value in unix time; so when we go to retrieve this from the cache, we just compare the current time (with a `Date.now()`) to the the retrieved value.

If the case is that we have NOT passed that retrieved unix time, then we will return the given errors, which are the same as the ones given for the original throttling from the server. If we have passed the retrieved unix time, we will send through the request like normal and clear the error/throttlingTime content from the cache.

It is a bit janky to not just have a separate key-value pair for erroneous cases, but we wanted to prioritize space in local storage. I'm willing to be convinced, though.

Reply:
Ah, I misread the code the first time. You fallback to Date.now() + const so as long as the retry-after header is the absolute unix time as well then it should be fine.